### ISSUE TITLE
Panic in decision_tree

### ISSUE THREAD

**Issue by insanitybit** [maciejkula/rustlearn#28]
Increasing the max_features in my hyperparameters seems to make this more likely.

```
let mut tree_params = decision_tree::Hyperparameters::new(feat_matrix.cols());
tree_params.min_samples_split(10)
           .rng(StdRng::from_seed(&[100]));

let mut model = Hyperparameters::new(tree_params, 20)
                    .rng(StdRng::from_seed(&[100]))
                    .one_vs_rest();
```

The data is shape:
rows: 2247
cols: 442

thread 'main' panicked at 'index out of bounds: the len is 10 but the index is 11', ../src/libcore/slice.rs:442

The len/index number changes, but it's always an off by one.

9:     0x55ede9c87a43 - core::panicking::panic_bounds_check::ha883fe1527ce6884
10:     0x55ede9a1e0a2 - _<[T] as core..slice..SliceExt>::swap::h9bf7fe18e2fa8251
                     at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcore/slice.rs:442
11:     0x55ede9a12fdc - collections::slice::_<impl [T]>::swap::h6559c8ef586608a3
                     at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/obj/../src/libcollections/slice.rs:479
12:     0x55ede9a27b9e - rustlearn::trees::decision_tree::FeatureIndices::mark_as_used::hc0584f311f525fd4
                     at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/rustlearn-0.4.0/src/trees/decision_tree.rs:103
13:     0x55ede9a298ae - rustlearn::trees::decision_tree::DecisionTree::build_tree::hdb922dd21da7568d

**Comment by maciejkula** [comment]
Thanks for the report, I'll look into it.

**Comment by maciejkula** [comment]
I added a tentative fix in #29.

You can test it by replacing you `Cargo.toml` entry for `rustlearn` with:

```
rustlearn = { git = "https://github.com/maciejkula/rustlearn.git", rev = "77b05b1717ce0bcf6070509956c251a9aa031baf" }
```

Please let me know if that fixes it!

**Comment by insanitybit** [comment]
This looks to have solved it - thank you.

**Comment by maciejkula** [comment]
The fix is now included in rustlearn 0.4.1 on crates.io. Thanks for reporting the bug!
